### PR TITLE
Add feature to get cluster configuration

### PR DESCRIPTION
**Pull request by quasiben**
Early attempt at resolving https://github.com/rapidsai/dask-cuda/issues/989

Screen Shot 2022-10-04 at 6 37 20 PM

In an ideal world, we could call `dask.config.config` or `dask.config.get(...)` and that will have been updated with live values.   Unfortunately, that is not the case.  Config options are set in all manner of places: cli, env, yaml, kwargs, etc and we are not consistent about storing them.  For example, in this PR `self.device_memory_limit = device_memory_limit` is added to `device_host_file.py` just so `client.run` could be accessed from within the worker.

This change employs a hodge-podge of techniques to get at options we _think_ may be potentially useful for debugging but in no way is this exhaustive.

### CODE REVIEW

File: dask_cuda/utils.py [hunk 1]
```diff
@@ -649,3 +649,101 @@ def get_gpu_uuid_from_index(device_index=0):
     pynvml.nvmlInit()
     handle = pynvml.nvmlDeviceGetHandleByIndex(device_index)
     return pynvml.nvmlDeviceGetUUID(handle).decode("utf-8")
+
+
+def get_worker_config(dask_worker):
+    import dask
+    from distributed.comm import parse_address
```
Comment (madsbk) [inline dask_cuda/utils.py:656]: Do we need the imports inside the function?

File: dask_cuda/utils.py [hunk 2]
```diff
@@ -649,3 +649,101 @@ def get_gpu_uuid_from_index(device_index=0):
...
+    for mem in [
+        "memory_limit",
+        "memory_pause_fraction",
+        "memory_spill_fraction",
+        "memory_target_fraction",
+    ]:
+        ret[mem] = getattr(dask_worker.memory_manager, mem)
+
+    # jit unspilling set
+    ret["jit-unspill"] = hasattr(dask_worker.data, "spill_on_demand_initialized")
```
Comment (madsbk) [inline dask_cuda/utils.py:679]: Maybe use `isinstance(dask_worker.data, ProxifyHostFile)` to make it clear what we consider JIT-Unspill

File: dask_cuda/device_host_file.py [hunk 3]
```diff
@@ -180,6 +180,8 @@ def __init__(
         local_directory=None,
         log_spilling=False,
     ):
+        # can now access value within dask-worker / client.run
+        self.device_memory_limit = device_memory_limit
```
Comment (madsbk) [inline dask_cuda/device_host_file.py:184]: Avoid this and use `data.device_buffer.n` to get device limit instead

File: dask_cuda/utils.py [hunk 4]
```diff
@@ -649,3 +649,101 @@ def get_gpu_uuid_from_index(device_index=0):
...
+
+    # get optional device-memory-limit
+    if ret["jit-unspill"]:
+        ret["device-memory-limit"] = dask_worker.data.manager._device_memory_limit
+    else:
+        device_memory_limit = getattr(dask_worker.data, "device_memory_limit", False)
```
Comment (madsbk) [inline dask_cuda/utils.py:685]: Use `dask_worker.data.device_buffer.n` to get device limit.

**Review by madsbk — COMMENTED** [review]
Looks good, I only have some minor suggestions

File: dask_cuda/utils.py [hunk 5]
```diff
@@ -649,3 +649,101 @@ def get_gpu_uuid_from_index(device_index=0):
...
+
+        table = Table(title="Dask Cluster Configuration")
+        table.add_column("Parameter", justify="left", style="red")
+        table.add_column("Value", justify="left", style="green")
+
+        params = sorted(ret.keys())
+        for p in params:
+            table.add_row(p, str(ret[p]))
+        console = Console()
+        console.print(table)
```
Comment (pentschev) [inline dask_cuda/utils.py:749]: Maybe make a CLI out of this? I can imagine this causing all sorts of issues if accidentally called from code when `rich` is installed, as it may try to rich-print on the console from arbitrary places.

**Review by pentschev — COMMENTED** [review]
Besides Mads suggestions, I have one more, but overall looks real nice @quasiben !

File: dask_cuda/utils.py (hunk 5 above)
Comment (quasiben) [reply dask_cuda/utils.py]: I added a CLI but left the rich code inside of `utils`.  i don't think it will cause any `rich-printing` .  My thought was that folks like @VibhuJawa / users may still want a nicely formatted table from within a live session

Comment (pentschev) [reply dask_cuda/utils.py]: > I added a CLI

I think a separate CLI (maybe `dask-get-config` or `dask-print-config`) would be better, given that has nothing to do with `dask-cuda-worker`.

> but left the rich code inside of `utils`. i don't think it will cause any `rich-printing` . My thought was that folks like @VibhuJawa / users may still want a nicely formatted table from within a live session

I agree a nicely formatted table is a nice addition to a live session, and maybe this is my ignorance of `rich` but I thought it was a tool to print nicely formatted tables on a terminal, am I mistaken? What if you call that code arbitrarily from any place in the code, say you want to parse that output and feed it to some other code, what happens then (assuming you have `rich` installed)?

Comment (VibhuJawa) [reply dask_cuda/utils.py]: > I added a CLI but left the rich code inside of utils. i don't think it will cause any rich-printing . My thought was that folks like @VibhuJawa / users may still want a nicely formatted table from within a live session

I like the nicely formatted output .   I guess if we want to feed it into some other code we just stick with the default of `Table=False` .

**Review by VibhuJawa — CHANGES_REQUESTED** [review]
Looks great to. 

Requesting a small change to make all the memory limits human readable as just parsing bytes is hard and this mainly for users to debug the cluster configuration.

File: dask_cuda/utils.py (hunk 5 above)
Comment (quasiben) [reply dask_cuda/utils.py]: Rich was, I think, originally designed for terminals but notebook support has existed for a 2 years or so:
https://github.com/Textualize/rich/issues/26

If you'd prefer a text only output I can look into the version mismatch output in dask:

```
+-------------+----------+-----------+----------+
| Package     | client   | scheduler | workers  |
+-------------+----------+-----------+----------+
| dask        | 2022.9.2 | 2022.7.1  | 2022.9.2 |
| distributed | 2022.9.2 | 2022.7.1  | 2022.9.2 |
+-------------+----------+-----------+----------+
```

@jacobtomlinson do you know if there are known issues with using rich in a jupyter session ?

File: dask_cuda/utils.py [hunk 6]
```diff
@@ -649,3 +653,105 @@ def get_gpu_uuid_from_index(device_index=0):
...
+    # device and host memory configuration
+    for p in plugin_vals:
+        for v in dir(p):
+            # ignore hidden attrs as well as setup and CPUAffinity
+            if v.startswith("_") or v == "setup" or v == "cores":
+                pass
+            else:
+                ret[v] = getattr(p, v)
+    for mem in [
+        "memory_limit",
```
[the remark this answers is not in the document]
Comment (quasiben) [reply dask_cuda/utils.py:672]: yup yup, i'll add that in

File: dask_cuda/utils.py [hunk 7]
```diff
@@ -649,3 +653,105 @@ def get_gpu_uuid_from_index(device_index=0):
...
+    ret = client.sync(
+        _get_cluster_configuration, client=client, asynchronous=client.asynchronous
+    )
+    if not table:
+        return ret
+    else:
+        try:
+            from rich.console import Console
+            from rich.table import Table
+        except ImportError:
```
Comment (wence-) [inline dask_cuda/utils.py:742]: I would advocate a stronger separation of acquiring the data from its presentation. That is, `get_cluster_configuration` should just return "raw" data, and there should be a separate function that prints a tabulated version.

File: dask_cuda/utils.py [hunk 8]
```diff
@@ -649,3 +653,105 @@ def get_gpu_uuid_from_index(device_index=0):
...
+            print(
+                "Please install rich `python -m pip install rich`"
+                / "to print a table of the current Dask Cluster Configuration"
+            )
+            sys.exit(1)
```
Comment (wence-) [inline dask_cuda/utils.py:747]: This is a very unfriendly way to complain about a missing package. Prefer `raise`ing some reasonable error here.

File: dask_cuda/utils.py [hunk 9]
```diff
@@ -649,3 +653,105 @@ def get_gpu_uuid_from_index(device_index=0):
...
+
+        table = Table(title="Dask Cluster Configuration")
+        table.add_column("Parameter", justify="left", style="red")
+        table.add_column("Value", justify="left", style="green")
+
+        params = sorted(ret.keys())
+        for p in params:
+            table.add_row(p, str(ret[p]))
```
Comment (wence-) [inline dask_cuda/utils.py:755]:
```suggestion
        for p in sorted(ret.keys()):
            table.add_row(p, str(ret[p]))
```
Or:
```
for key, value in sorted(ret.items(), key=operator.itemgetter(0)):
    table.add_row(key, str(value))
```
If you like

Comment (wence-) [reply dask_cuda/utils.py:755]: Aside: why do we need to explicit `str` the value, but not the key? With my proposed change to providing nested dictionaries for providing configuration settings of plugins would suggest one could format as nested tables.

File: dask_cuda/utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (wence-) [reply dask_cuda/utils.py:672]: Please prefer doing this in the presentation, rather than data-gathering, phase.

File: dask_cuda/tests/test_local_cuda_cluster.py [hunk 10]
```diff
@@ -350,7 +355,10 @@ async def test_gpu_uuid():
 async def test_rmm_track_allocations():
     rmm = pytest.importorskip("rmm")
     async with LocalCUDACluster(
-        rmm_pool_size="2GB", asynchronous=True, rmm_track_allocations=True
+        rmm_pool_size="2GB",
+        asynchronous=True,
+        rmm_track_allocations=True,
+        device_memory_limit="30B",
```
Comment (wence-) [inline dask_cuda/tests/test_local_cuda_cluster.py:361]: Why this change here?

File: dask_cuda/utils.py [hunk 11]
```diff
@@ -649,3 +653,105 @@ def get_gpu_uuid_from_index(device_index=0):
     pynvml.nvmlInit()
     handle = pynvml.nvmlDeviceGetHandleByIndex(device_index)
     return pynvml.nvmlDeviceGetUUID(handle).decode("utf-8")
+
+
+def get_worker_config(dask_worker):
```
Comment (wence-) [inline dask_cuda/utils.py:658]: Why is this function not private (`_get_worker_config`)? Or, conversely, why are the other functions not public?

File: dask_cuda/utils.py [hunk 12]
```diff
@@ -649,3 +653,105 @@ def get_gpu_uuid_from_index(device_index=0):
...
+    ret = {}
+
+    # device and host memory configuration
+    for p in plugin_vals:
+        for v in dir(p):
+            # ignore hidden attrs as well as setup and CPUAffinity
+            if v.startswith("_") or v == "setup" or v == "cores":
+                pass
+            else:
+                ret[v] = getattr(p, v)
```
Comment (wence-) [inline dask_cuda/utils.py:670]: This assumes that all plugins have disjoint attributes.

```
plugin_config = {}
for p in plugin_vals:
    plugin_config[type(p).__name__] = {v: getattr(p, v) for v in dir(p) if not (v.startswith("_") or v in {"setup", "cores"})}
```
(Untested)

File: dask_cuda/utils.py [hunk 13]
```diff
@@ -649,3 +653,105 @@ def get_gpu_uuid_from_index(device_index=0):
...
+
+        table = Table(title="Dask Cluster Configuration")
+        table.add_column("Parameter", justify="left", style="red")
+        table.add_column("Value", justify="left", style="green")
+
+        params = sorted(ret.keys())
+        for p in params:
+            table.add_row(p, str(ret[p]))
+        console = Console()
+        console.print(table)
```
Comment (wence-) [inline dask_cuda/utils.py:757]:
```suggestion
        Console().print(table)
```

**Review by wence- — CHANGES_REQUESTED** [review]
Some nitpicks

File: dask_cuda/utils.py [hunk 14]
```diff
@@ -649,3 +653,105 @@ def get_gpu_uuid_from_index(device_index=0):
...
+        except ImportError:
+            print(
+                "Please install rich `python -m pip install rich`"
+                / "to print a table of the current Dask Cluster Configuration"
+            )
+            sys.exit(1)
+
+        table = Table(title="Dask Cluster Configuration")
+        table.add_column("Parameter", justify="left", style="red")
+        table.add_column("Value", justify="left", style="green")
```
Comment (wence-) [inline dask_cuda/utils.py:751]: Should we really be opinionated about colours here? (I noticed in your screenshot that red on black was not very readable).

Comment (quasiben) [reply dask_cuda/utils.py:751]: Yes!  I would love direction on styling.  If you have a specific set you think works well I'm happy to change

File: dask_cuda/utils.py (hunk 9 above)
Comment (quasiben) [reply dask_cuda/utils.py:755]: Rich uses [renderables](https://rich.readthedocs.io/en/latest/protocol.html#console-render) which are either strings or have implemented valid `__rich__` / `__rich_console__` methods:
https://github.com/Textualize/rich/blob/84e628655a2981ee90413ca3f35001ec3954161d/rich/protocol.py#L10-L16

File: dask_cuda/tests/test_local_cuda_cluster.py (hunk 10 above)
Comment (quasiben) [reply dask_cuda/tests/test_local_cuda_cluster.py:361]: This was an accident while I was mucking about with config tests.

File: dask_cuda/utils.py (hunk 14 above)
Comment (wence-) [reply dask_cuda/utils.py:751]: I'm not sure I do have a set that works. We could use the styles that rich defines, rather than specific colours e.g. "bold" (see them with `python -m rich.default_styles`) and then if someone has overridden the styling with a configuration file that will be obeyed.

File: dask_cuda/utils.py (hunk 11 above)
Comment (quasiben) [reply dask_cuda/utils.py:658]: I'm following a pattern used in dask where the fully async function definition is private and the public function is both sync/async.  For example, we do something similar for [all_to_all](https://github.com/rapidsai/dask-cuda/blob/23460f6f7300202a2d3a25ffedf6884755fe6fbc/dask_cuda/utils.py#L444-L470)

File: dask_cuda/utils.py (hunk 12 above)
Comment (quasiben) [reply dask_cuda/utils.py:670]: This is a lot clean and works nicely.  Thank you!

Comment (quasiben) [reply dask_cuda/utils.py:670]: Hmm, on second thought, I think it's better to key something like I had before but flatten the key name like:

> key = type(p).__name__ + '.' + v

This makes it easier to read without having to scan a nested dictionary within the table

**Comment by quasiben** [comment]
Thanks for the review comments @wence-  .  I believe it's ready for another round.  As for styling, what do you think about bold bright_green ?

Screen Shot 2022-10-11 at 10 08 05 AM

File: dask_cuda/utils.py (hunk 12 above)
Comment (wence-) [reply dask_cuda/utils.py]: > This makes it easier to read without having to scan a nested dictionary within the table

Or one could translate (in the rich output), like so:

```python
from functools import singledispatch
from rich.pretty import Pretty
from rich.table import Table
from operator import itemgetter

@singledispatch
def pretty(obj):
    return Pretty(obj)

@pretty.register(dict)
def pretty_dict(obj):
    t = Table()
    t.add_column("key")
    t.add_column("value")
    for k, v in sorted(obj.items(), key=itemgetter(0)):
        t.add_row(pretty(k), pretty(v))
    return t

...

data = get_config(...)
Console().print(pretty(data))
```

WDYT?

File: dask_cuda/cli/dask_cuda_worker.py [hunk 15]
```diff
@@ -321,8 +333,14 @@ def main(
...
     pre_import,
     multiprocessing_method,
+    get_cluster_conf,
     **kwargs,
 ):
+    if get_cluster_conf:
+        client = Client(scheduler)
+        print_cluster_config(client)
+        sys.exit(0)
+
```
Comment (wence-) [inline dask_cuda/cli/dask_cuda_worker.py:343]: This needs to move further down (and respect `scheduler_file` and `security` arguments).

File: dask_cuda/utils.py [hunk 16]
```diff
@@ -649,3 +653,116 @@ def get_gpu_uuid_from_index(device_index=0):
...
+        ret["ucx-transports"] = ucp.get_active_transports()
+
+    # comm timeouts
+    ret["distributed.comm.timeouts"] = dask.config.get("distributed.comm.timeouts")
+    # ret.update(plugin_config)
+
+    return ret
+
+
+async def _get_sched_config(client):
```
Comment (wence-) [inline dask_cuda/utils.py:705]: nit: consistent naming? `_get_scheduler_config`; and then perhaps, given we've called it `get_cluster_configuration`, `_get_scheduler_configuration` and `get_worker_configuration` ?

File: dask_cuda/utils.py [hunk 17]
```diff
@@ -649,3 +653,116 @@ def get_gpu_uuid_from_index(device_index=0):
...
+def print_cluster_config(client):
+    """print current Dask cluster configuration"""
+    data = get_cluster_configuration(client)
+    formatted_byte_keys = [
+        "memory_limit",
+        "device-memory-limit",
+        "initial_pool_size",
+        "maximum_pool_size",
+    ]
+
```
Comment (wence-) [inline dask_cuda/utils.py:742]: Should we add the size of the cluster? (Number of workers/theads/gpus/etc...)

**Review by wence- — COMMENTED** [review]
Minor

**Comment by wence-** [comment]
> ```diff
> diff --git a/dask_cuda/cli/dask_cuda_worker.py b/dask_cuda/cli/dask_cuda_worker.py
> index 464d988..92b61e8 100755
> --- a/dask_cuda/cli/dask_cuda_worker.py
> +++ b/dask_cuda/cli/dask_cuda_worker.py
> @@ -336,11 +336,6 @@ def main(
>      get_cluster_conf,
>      **kwargs,
>  ):
> -    if get_cluster_conf:
> -        client = Client(scheduler)
> -        print_cluster_config(client)
> -        sys.exit(0)
> -
>      if multiprocessing_method == "forkserver":
>          import multiprocessing.forkserver as f
>  
> @@ -361,6 +356,14 @@ def main(
>              "unsupported one. Scheduler address: %s" % scheduler
>          )
>  
> +    if get_cluster_conf:
> +        if scheduler_file is not None:
> +            client = Client(scheduler_file=scheduler_file, security=security)
> +        else:
> +            client = Client(scheduler, security=security)
> +        print_cluster_config(client)
> +        sys.exit(0)
> +
>      if worker_class is not None:
>          worker_class = import_term(worker_class)
> ```

This hunk is needed irrespective of the proposed formatting changes.

**Comment by quasiben** [comment]
@wence- I'll your suggestion a lot. Made a minor adjustment (need to default to `None` for `--scheduler-file`.  I'll push changes shortly with the other suggestions as well

File: dask_cuda/utils.py (hunk 12 above)
Comment (quasiben) [reply dask_cuda/utils.py]: 👍  - this is really cool!

File: dask_cuda/utils.py [hunk 18]
```diff
@@ -649,3 +653,116 @@ def get_gpu_uuid_from_index(device_index=0):
...
+            value = format_bytes(value)
+        table.add_row(key, str(value))
+    Console().print(table)
+
+
+def get_cluster_configuration(client):
+    data = client.sync(
+        _get_cluster_configuration, client=client, asynchronous=client.asynchronous
+    )
+    return data
```
[the remark this answers is not in the document]
Comment (quasiben) [reply dask_cuda/utils.py:768]: Actually, I think we need to go the other way and build an async print function.  This would allow of for usage of `print_cluster_config` inside a jupyter notebook with an async client

File: dask_cuda/utils.py [hunk 19]
```diff
@@ -649,3 +654,144 @@ def get_gpu_uuid_from_index(device_index=0):
...
+
+
+def get_worker_config(dask_worker):
+    # assume homogenous cluster
+    plugin_vals = dask_worker.plugins.values()
+    ret = {}
+
+    # device and host memory configuration
+    for p in plugin_vals:
+        ret[type(p).__name__] = {
```
Comment (wence-) [inline dask_cuda/utils.py:666]: Looking again (sorry!), shall we indicate that these are plugins? `ret[f"[plugin] {type(p).__name__}"] = ...` ?

**Comment by wence-** [comment]
Thanks @quasiben, I'm pretty happy with the output, and we can iterate separately if we need to afterwards. I think the outstanding queries to address here are @pentschev's comment about whether this really makes sense as an additional flag to `dask-cuda-worker` or whether it should be a separate commandline tool. We've now separated the data gathering from printing, so I think a separate commandline tool makes more sense.

This would also fix the likelihood of making the error that I did when trying to use it which was to try and launch a new worker and print the cluster configuration. This new option is really "connect to an existing cluster and dump information".

WDYT?

File: dask_cuda/utils.py (hunk 19 above)
Comment (quasiben) [reply dask_cuda/utils.py:666]: I'm happy to include it and we can continue iterating.  I doubt most users deeply care about what is label a plugin but we (developers/anyone helping to debug) may find it useful.

**Comment by quasiben** [comment]
I'll build a new CLI: `dask-config` this is also related to an issue I'm going to file today around configuring "profiles" for dask/RAPIDS

**Comment by quasiben** [comment]
@wence- when you have some time can you give this PR another review ?

File: dask_cuda/cli/dask_config.py [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+
+
+pem_file_option_type = click.Path(exists=True, resolve_path=True)
+
+
+@click.command(context_settings=dict(ignore_unknown_options=True))
+@click.argument("scheduler", type=str, required=False)
+@click.argument(
+    "preload_argv", nargs=-1, type=click.UNPROCESSED, callback=validate_preload_argv
+)
```
Comment (wence-) [inline dask_cuda/cli/dask_config.py:24]: These arguments don't seem to go anywhere? [...], oh I see, `validate_preload_argv` side-effectfully imports the specified module. Do we need this here?

File: dask_cuda/cli/dask_config.py [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+            "unsupported one. Scheduler address: %s" % scheduler
+        )
+
+    if get_cluster_conf:
+        if scheduler_file is not None:
+            client = Client(scheduler_file=scheduler_file, security=security)
+        else:
+            client = Client(scheduler, security=security)
+        print_cluster_config(client)
+        sys.exit(0)
```
Comment (wence-) [inline dask_cuda/cli/dask_config.py:93]: No need to `sys.exit` here since this is the end of the function.
```suggestion
```

File: dask_cuda/cli/dask_config.py [hunk 22]
```diff
@@ -0,0 +1,101 @@
...
+
+
+def go():
+    main()
+
+
+if __name__ == "__main__":
+    go()
```
Comment (wence-) [inline dask_cuda/cli/dask_config.py:101]:
```suggestion
if __name__ == "__main__":
    main()
```

?

File: dask_cuda/utils.py [hunk 23]
```diff
@@ -649,3 +654,147 @@ def get_gpu_uuid_from_index(device_index=0):
...
+    scheme, loc = parse_address(dask_worker.scheduler.address)
+    ret["protocol"] = scheme
+    if scheme == "ucx":
+        import ucp
+
+        ret["ucx-transports"] = ucp.get_active_transports()
+
+    # comm timeouts
+    ret["distributed.comm.timeouts"] = dask.config.get("distributed.comm.timeouts")
+    # ret.update(plugin_config)
```
Comment (wence-) [inline dask_cuda/utils.py:701]:
```suggestion
```

File: dask_cuda/utils.py [hunk 24]
```diff
@@ -649,3 +654,147 @@ def get_gpu_uuid_from_index(device_index=0):
...
+@singledispatch
+def pretty_print(obj, toplevel):
+    from rich.pretty import Pretty
+
+    return Pretty(obj)
+
+
+@pretty_print.register(str)
+def pretty_print_str(obj, toplevel):
+    return obj
```
Comment (wence-) [inline dask_cuda/utils.py:747]:
```suggestion
    from rich.markup import escape
    return escape(obj)
```

Moving the escaping to `str`s

File: dask_cuda/utils.py [hunk 25]
```diff
@@ -649,3 +654,147 @@ def get_gpu_uuid_from_index(device_index=0):
...
+
+
+@pretty_print.register(dict)
+def pretty_print_dict(obj, toplevel):
+    from rich.markup import escape
```
Comment (wence-) [inline dask_cuda/utils.py:752]:
```suggestion
```

**Review by wence- — CHANGES_REQUESTED** [review]
Very minor nits, but otherwise looks great! I like the output.

**Comment by quasiben** [comment]
Thanks @wence- for the additional comments!

**Comment by wence-** [comment]
Think you need a little bit more linting?

**Review by wence- — APPROVED** [review]
Thanks!

**Comment by quasiben** [comment]
Thanks again for the comments and review @wence- @pentschev @madsbk

**Comment by quasiben** [comment]
@gpucibot merge
